Add ApplicationService Mounting

joshdover opened the pull request:
## Summary

Implements [RFC-0004](https://github.com/elastic/kibana/blob/master/rfcs/text/0004_application_service_mounting.md)
Related to #18843
Depends on #41251 

This adds support for UI applications in the New Platform. With this change, CoreSystem on the frontend now has a "legacy mode" and normal mode.
- In legacy mode, the ApplicationService does not mount its router on the page and instead gives control to the LegacyService.
- In normal mode, the ApplicationService mounts its router on the page and controls mounting individual UI apps. The LegacyService does import any legacy code, though it does still provide some data and features to mirror Legacy behaviors.

This mode is toggled by the legacy `renderApp` function, which serves the bootstrapping page and bundles. When rendering a UI app that is unknown to the legacy platform, this function will instead serve the "core-only" bundle which does not include any legacy code or applications.

New Platform applications currently have some caveats:
- Telemetry banner will not show (need #41986 implemented)
- No legacy code, including uiExport "hacks" are run when NP apps are running.
  - This means some features will not work on these pages, such as reporting's polling for completed jobs.
  - I've done an audit of all our uiExport hacks that need to be updated: #42984 
- ~~Nav links for NP apps do not support "open in new window". To support this, we need EUI to support links with both `href` and `onClick` props (elastic/eui#1933)~~

### Testing

To test the NP application service yourself, [unzip these plugins](https://github.com/elastic/kibana/files/3493287/plugins.zip) into your `plugins` directory, run `yarn kbn bootstrap`, and start Kibana.

Alternatively, you can also start the functional test server with the config that loads some test applications:

```
node scripts/functional_tests_server.js --config test/plugin_functional/config.js
```

### Dev Docs

Kibana Platform plugins may now register UI applications to display in the global navigation and be mounted inside Kibana as a single-page application. Example:

```ts
class MyPlugin {
  public setup(core) {
    core.application.register({
      id: 'myApp',
      async mount(context, params) {
        const { renderApp } = await import('./application');
        return renderApp(context, params);
      }
    });
  }
}
```

For more details, see the [Application Mounting RFC](https://github.com/elastic/kibana/blob/master/rfcs/text/0004_application_service_mounting.md).

### Checklist

Use ~~strikethroughs~~ to remove checklist items you don't feel are applicable to this PR.

- [x] This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)
- [x] [Documentation](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#writing-documentation) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios

### For maintainers

- [x] This was checked for breaking API changes and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)
- [x] This includes a feature addition or change that requires a release note and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)

elasticmachine commented:
Pinging @elastic/kibana-platform

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7136/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7241/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7820/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9227/)

joshdover committed:
Add core-only bundle

joshdover committed:
Add ApplicationService mounting

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9851/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10071/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10122/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10535/)

joshdover commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10653/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10666/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10694/)

joshdover committed:
Add LegacyCore{Setup,Start}

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11126/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11184/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11345/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11378/)

restrry commented:
ack: going to review tomorrow

jbudz reviewed (approved):
👍 renovate.json

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11594/)

streamich commented on src/core/public/application/types.ts:
```diff
@@ -0,0 +1,293 @@
+  /**
+   * Core service APIs available to mounted applications.
+   */
+  core: {
```
Is this different from `CoreStart`? Maybe it could be

```ts
core: CoreStart;
```

joshdover replied:
It is the same currently, but that will change in the very near future. For instance, part of this context will include an API for blocking navigation to other apps that will not be available in CoreStart but will be available in `context.core.application.blockNavigation`

restrry commented on src/legacy/ui/public/new_platform/new_platform.ts:
```diff
@@ -34,32 +34,32 @@ export interface PluginsStart {
 /**
  * Only used by unit tests
- * @internal
+ * @Legacy
```
unnecessary change?

restrry commented on src/legacy/ui/ui_render/ui_render_mixin.js:
```diff
@@ -198,16 +203,15 @@ export function uiRenderMixin(kbnServer, server, config) {
     const request = h.request;
     const basePath = request.getBasePath();
+    const uiSettings = await getUiSettings({ request, includeUserProvidedConfig });
+    app = app || { getId: () => 'core' };
```
shouldn't we cover it with unit/integration tests?

joshdover replied:
Good call, I'll add some tests for this file.

mshustov reviewed (commented):
don't have time to play with implementation. continue on Monday

restrry commented:
> To test the NP application service yourself, unzip these plugins into your plugins directory, run yarn kbn bootstrap, and start Kibana.

it contains `foo` and `bar` plugins not compatible with the latest implementation. Do you think you could update them to run locally?
don't see functional tests either to play around with.

joshdover commented on src/core/public/core_system.ts:
```diff
@@ -200,6 +223,17 @@ export class CoreSystem {
       });
       const uiSettings = await this.uiSettings.start();
 
+      application.registerMountContext(this.coreContext.coreId, 'core', () => ({
```
I don't think anything during `setup` should be callable during any handlers.

For the fatalErrors example, I plan to not expose this to plugins at all and replace it with an ApplicationService-level API that only renders UI in the application container rather than over the entire page: https://github.com/elastic/kibana/issues/41980

Plugins can still create a fatal error by throwing an error during their `setup` or `start` functions, which will automatically be handled by Core.

joshdover commented:
> it contains `foo` and `bar` plugins not compatible with the latest implementation. Do you think you could update them to run locally?

I've updated the plugins zip. There were some type problems that I fixed.

> don't see functional tests either to play around with.

I'm working on adding some this week, but wanted to put this up for review sooner than later. I should have noted that in the description 😅

joshdover committed:
Fix PR comments

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12025/)

joshdover committed:
Add functional tests

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12050/)

restrry commented on src/core/public/application/types.ts:
```diff
@@ -252,7 +255,7 @@ export interface ApplicationStart {
    * @param provider - A {@link IContextProvider} function
    */
-  registerMountContext<T extends keyof AppMountContext>(
+  registerMountContext<T extends AppMountContextNames>(
```
We shouldn't utilize once calculated types for extendable types.
![2019-08-13_12-18-01](https://user-images.githubusercontent.com/3198181/62934383-b4702980-bdc4-11e9-8f4d-d50e4166034d.png)

restrry commented on test/plugin_functional/test_suites/core_plugins/applications.js:
```diff
@@ -0,0 +1,82 @@
+      // Go to page A
+      await testSubjects.click('foo-nav-page-a');
+      expect(await browser.getCurrentUrl()).to.eql(`http://localhost:5620/app/foo/page-a`);
+      await testSubjects.existOrFail('foo-app-page-a');
```
should we check that page was not reloaded?

restrry commented on src/core/public/application/ui/ui_integration.test.tsx:
```diff
@@ -0,0 +1,131 @@
+/*
```
nit: integration tests are resided in `integration_test` folder

restrry commented on src/core/public/rendering/rendering_service.tsx:
```diff
@@ -39,28 +43,40 @@ interface StartDeps {
 
-          <div ref={legacyRef} />
+          {!legacyMode && (
+            <div className="app-wrapper">
```
should we add a test to check `app-wrapper` is present in DOM tree? it's a part of public API https://github.com/restrry/kibana/blob/fe51d3419fefc464e8a5672d66bb625d7d5e6b06/x-pack/legacy/plugins/canvas/public/components/popover/popover.js#L67

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12121/)

restrry commented on src/legacy/ui/ui_render/ui_render_mixin.js:
```diff
@@ -164,12 +164,11 @@ export function uiRenderMixin(kbnServer, server, config) {
 
   server.route({
-    path: '/app/{id}',
+    path: '/app/{id}/{path?}',
```
does anyone read this parameter?

joshdover replied:
No, but it's necessary for Hapi to match SPA routes like `/app/foo/deep/path`

restrry replied:
I see. optional: for consistency with other code `'/app/{id}/{p*}',
[1](https://github.com/restrry/kibana/blob/fe51d3419fefc464e8a5672d66bb625d7d5e6b06/src/legacy/server/http/index.js#L98)
[2](https://github.com/restrry/kibana/blob/fe51d3419fefc464e8a5672d66bb625d7d5e6b06/x-pack/legacy/plugins/code/server/routes/redirect.ts#L29)
[3](https://github.com/restrry/kibana/blob/fe51d3419fefc464e8a5672d66bb625d7d5e6b06/x-pack/legacy/plugins/reporting/server/routes/generate.ts#L45)

joshdover commented:
Taking off for the week 🏖, I'll update this when I get back!

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12281/)

joshdover committed:
Fix PR comments

joshdover commented:
@restrry Addressed all comments!

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13148/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13158/)

eliperelman commented on src/core/public/application/ui/app_container.tsx:
```diff
@@ -0,0 +1,114 @@
+  private readonly containerDiv = React.createRef<HTMLDivElement>();
+  private unmountFunc?: AppUnmount;
+
+  constructor(props: Props) {
```
I think this can be shortened to a class field:

```diff
- constructor(props: Props) {
-   super(props);
-   this.state = { appNotFound: false };
- }
+ state = { appNotFound: false };
```

eliperelman commented on src/core/public/application/ui/app_container.tsx:
```diff
@@ -0,0 +1,114 @@
+
+  async mountApp() {
+    const { apps, legacyApps, match, basePath, currentAppId$, redirectTo } = this.props;
+    const appId = match.params.appId;
```
Can be deconstructed:

```js
const { appId } = match.params;
```

eliperelman commented on src/core/public/chrome/nav_links/nav_links_service.ts:
```diff
@@ -99,10 +99,22 @@ export class NavLinksService {
   public start({ application, http }: StartDeps): ChromeNavLinks {
-    const legacyAppLinks = application.availableLegacyApps.map(
-      app =>
+    const appLinks = [...application.availableApps.entries()].map(
```
nit: `.entries()` is redundant.

eliperelman commented on src/core/public/chrome/nav_links/nav_links_service.ts:
```diff
@@ -99,10 +99,22 @@ export class NavLinksService {
+        ] as [string, NavLinkWrapper]
+    );
+
+    const legacyAppLinks = [...application.availableLegacyApps.entries()].map(
```
nit: `.entries()` is redundant.

eliperelman commented on src/core/public/chrome/ui/header/header.tsx:
```diff
@@ -287,9 +299,18 @@ class HeaderUI extends Component<Props, State> {
+        // This needs to work with both href and onClick to support "open in new tab" correctly, however EUI
+        // does not current support this.
+        // https://github.com/elastic/eui/pull/1933
+        href: legacyMode || navLink.legacy ? navLink.href : undefined,
```
We can simplify this to only check the condition once via:

```ts
...(legacyMode || navLink.legacy
  ? { href: navLink.href }
  : { onClick: () => navigateToApp(navLink.id) }
),
```

joshdover replied:
I plan on fixing the caveat mentioned in the comment once EUI is upgraded. That should happen before this merges, so I'm just going to leave this as-is since it will change anyways.

eliperelman commented on src/legacy/ui/public/chrome/chrome.js:
```diff
@@ -95,7 +95,6 @@ const waitForBootstrap = new Promise(resolve => {
     document.body.setAttribute('id', `${internals.app.id}-app`);
 
     chrome.setupAngular();
-    // targetDomElement.setAttribute('id', 'kibana-body');
```
Extraneous comment.

joshdover replied:
...yep that's what I deleted it 😄

eliperelman commented on test/functional/page_objects/common_page.js:
```diff
@@ -151,11 +151,21 @@ export function CommonPageProvider({ getService, getPageObjects }) {
-        hash: hash || appConfig.hash,
-      });
+
+      let appUrl;
```
Can be compacted via ternary to:

```ts
const appConfig = config.get(['apps', appName]);
const appUrl = getUrl.noAuth(
  config.get('servers.kibana'),
  appConfig
    ? { pathname: `${basePath}${appConfig.pathname}`, hash: hash || appConfig.hash }
    : { pathname: `${basePath}/app/${appName}` }
);
```

joshdover replied:
This doesn't work because `config.get(['apps', appName])` throws an exception if it doesn't exist.

eliperelman replied:
Aww, that's unfortunate.

eliperelman reviewed (approved):
LGTM, most of the comments are just nits. :+1:

joshdover commented on src/core/public/application/application_service.test.tsx:
```diff
@@ -17,57 +17,219 @@
-  },
-]
-`);
+                        Map {
```
When Jest writes a new snapshot to the file it messes these up sometimes, it kept happening so I stopped fighting it.

rudolf replied:
I also had a couple of fights with jest about this ;) it's a known issue: https://github.com/facebook/jest/issues/8424

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13601/)

rudolf commented on src/core/public/application/types.ts:
```diff
@@ -0,0 +1,293 @@
+  /**
+   * An ordinal used to sort nav links relative to one another for display.
+   */
+  order?: number;
```
Why would we want `order` to be optional? If two or more apps don't specify an ordering then the ordering might not be guaranteed to be stable. I'm guessing if a non-unique order number is specified the ordering is affected by some underlying implementation detail like the order in which apps were registered, which in turn could be affected by the dependency graph?

joshdover replied:
Correct, the ordering is currently stable by virtue of the plugin discovery's topological order algorithm. In practice, this means new apps tend to show up at the bottom of the list.

I'm not too keen on making plugin devs (mostly 3rd party ones) specify this field as it feels like just another detail to fill in that is unimportant.

rudolf replied:
Yeah, I agree, from the perspective of 3rd party plugins it doesn't make sense to choose an ordering number.

joshdover committed:
Fix PR comments

joshdover commented on src/core/public/application/types.ts:
```diff
@@ -0,0 +1,293 @@
+ */
+export interface App extends AppBase {
+  /**
+   * A mount function called when the user navigates to this app's `rootRoute`.
```
Yep, fixed!

rudolf replied:
Seem like `rootRoute` slipped back into the changes.

joshdover commented on rfcs/text/0004_application_service_mounting.md:
```diff
@@ -109,11 +114,11 @@ export interface AppSpec {
    * @returns An unmounting function that will be called to unmount the application.
    */
-  mount(context: MountContext, targetDomElement: HTMLElement): Unmount | Promise<Unmount>;
+  mount(context: MountContext, params: AppMountParams): Unmount | Promise<Unmount>;
```
The `context` is values provided by other services that may or may not be relevant to an app's `mount` function. The `params` argument is the values provided by the ApplicationService that you need to mount your app.

Similar to how HTTP provides multiple arguments to route handlers (`(context, request, response)`), the ApplicationService does too. I think this makes it more explicit that these values are important to the mounting routine. I could make these more consistent with HTTP by providing them as separate arguments. What do you think?

rudolf replied:
Ah OK, yes the explanation and analogy with HTTP route handlers makes a lot of sense. I don't think it's that important that it mirrors the http route handler interface with the separate arguments, so don't have a strong preference either way.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14116/)

joshdover committed:
Remove other usages of rootRoute

joshdover committed:
Merge remote-tracking branch 'upstream/master' into app-service/mount

joshdover committed:
Use state field notation

joshdover committed:
Merge remote-tracking branch 'upstream/master' into app-service/mount

joshdover committed:
Add support for open in new tab

restrry commented on test/plugin_functional/plugins/core_plugin_b/public/plugin.tsx:
```diff
@@ -34,6 +35,15 @@ export class CorePluginBPlugin
+    core.application.register({
+      id: 'bar',
+      title: 'Bar',
+      async mount(context, params) {
```
we need to create recommendation how to declare additional properties on `context`. if we delete this import https://github.com/elastic/kibana/pull/41007/files#diff-73111fdf9de323e24d3cf6695ad53aafR21 `context` doesn't contain methods provided by dependency. because dependency declaration is not imported 
```ts
declare module 'kibana/public' {
  interface AppMountContext {...}
}
```

restrry commented on package.json:
```diff
@@ -119,6 +119,7 @@
     "@kbn/pm": "1.0.0",
     "@kbn/test-subj-selector": "0.2.1",
     "@kbn/ui-framework": "1.0.0",
+    "@types/history": "^4.7.2",
```
it's a dev dependency

restrry commented on src/core/public/application/application_service.test.tsx:
```diff
@@ -17,57 +17,219 @@
+    injectedMetadata.getLegacyMode.mockReturnValue(false);
+    const start = await service.start({ http, injectedMetadata });
+
+    expect(shallow(React.createElement(() => start.getComponent()))).toMatchSnapshot();
```
Is there a way to check a method returns a JSX tree, without testing a tree content? The current test may lead to false-positive fails.

restrry commented on src/core/public/application/application_service.tsx:
```diff
@@ -17,140 +17,166 @@
+      registerMountContext: this.mountContext.registerContext,
+      currentAppId$,
+
+      urlForApp: (appId, options: { path?: string } = {}) => {
```
optional: usually a method name starts with a verb. `urlForApp` -> `getUrlForApp`.

restrry commented on src/core/public/http/http_service.mock.ts:
```diff
@@ -25,7 +25,7 @@ type ServiceSetupMockType = jest.Mocked<HttpSetup> & {
 };
 
-const createServiceMock = (): ServiceSetupMockType => ({
+const createServiceMock = (basePath = ''): ServiceSetupMockType => ({
```
let's add it as a named argument. I suspect most of them are will be optional

restrry commented on src/core/public/index.ts:
```diff
@@ -156,15 +161,34 @@ export interface CoreStart {
+ * @public
+ * @deprecated
+ */
+// eslint-disable-next-line @typescript-eslint/no-empty-interface
```
unnecessary comment

restrry commented on src/core/public/plugins/plugin_context.ts:
```diff
@@ -76,6 +76,11 @@ export function createPluginSetupContext<
+      registerMountContext: (contextName, provider) =>
+        deps.application.registerMountContext(plugin.opaqueId, contextName, provider),
+    },
     context: omit(deps.context, 'setCurrentPlugin'),
```
there is no `setCurrentPlugin` in context. is there?

restrry commented on src/core/public/plugins/plugins_service.test.ts:
```diff
@@ -97,9 +98,7 @@ beforeEach(() => {
-    application: {
-      capabilities: mockStartDeps.application.capabilities,
-    },
+    application: expect.any(Object),
```
what we want to test here?

restrry replied:
1. I'm not sure assertion works outside of expect. https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/jest/index.d.ts#L547-L563 this test uses it when creates a new object
2. In other places when we test interfaces it's excessive, because TS performs the same check

restrry commented on src/core/public/rendering/rendering_service.test.tsx:
```diff
@@ -21,46 +21,74 @@ import React from 'react';
+      `);
+    });
+
+    it('returns a div for the legacy service to render into', () => {
```
suit name a bit confusing, we don't test target element here, but its children

restrry commented on test/functional/page_objects/common_page.js:
```diff
@@ -151,11 +151,21 @@ export function CommonPageProvider({ getService, getPageObjects }) {
+        });
+      } else {
+        appUrl = getUrl.noAuth(config.get('servers.kibana'), {
+          pathname: `${basePath}/app/${appName}`
```
shouldn't we pass `hash` for compatibility with old version?

restrry commented on test/plugin_functional/plugins/core_plugin_a/public/plugin.tsx:
```diff
@@ -18,9 +18,19 @@
+      id: 'foo',
+      title: 'Foo',
+      async mount(context, params) {
+        // Once dynamic imports are supported, replace with `await import()`
```
already merged

joshdover committed:
Merge remote-tracking branch 'upstream/master' into app-service/mount

joshdover committed:
Fix PR comments

restrry commented on src/core/public/application/ui/app_container.tsx:
```diff
@@ -0,0 +1,111 @@
+
+  async unmountApp() {
+    if (this.unmountFunc) {
+      await this.unmountFunc();
```
seems that commit when I removed `async/await` was rewritten. could you make the method sync again?

restrry commented on src/core/public/chrome/ui/header/header.tsx:
```diff
@@ -115,13 +116,24 @@ function extendRecentlyAccessedHistoryItem(
   };
 }
 
+function isModifiedEvent(event: MouseEvent) {
```
no tests. I'm not sure that our testing framework supports `keyDown + click` tho

joshdover committed:
Merge remote-tracking branch 'upstream/master' into app-service/mount

joshdover committed:
Fix pesky await from the dead

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/317/)

joshdover committed:
Merge remote-tracking branch 'upstream/master' into app-service/mount

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/497/)

joshdover committed:
Update docs

joshdover committed:
Merge remote-tracking branch 'upstream/master' into app-service/mount

joshdover committed:
Bump @types/history

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/770/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/804/)

rayafratkina commented:
@joshdover can you please add a DevDocs section?

joshdover commented:
@rayafratkina Done.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/9636/)
* Commit: 4152e92be5ca3c07002609f7eefcfd131606de56
